docs/investigations/gateway-architecture/gateway-architecture.md

```diff
@@ -0,0 +1,120 @@
+2. Stitching multiple GraphQL schemas written for internal services into a single one and proxying the traffic to those internal GraphQL servers
+3. Having separate HTTP endpoints proxying traffic to specific internal servers (exposing many separate GraphQL APIs)
+
+## Possible solutions
+
+### 1. Handwritten GraphQL Gateway
+
+The first option is to use a single GraphQL server on the gateway with manually written schema and resolvers, that would delegate the operations to internal services. We could use for example gRPC to communicate with them (a PoC of this can be found [here](https://github.com/kyma-incubator/compass/pull/21/)).
+
+Pros
+
+- full GraphQL support (introspection + subscriptions)
+- avoiding Node.js (better performance)
+- single endpoint
+- facade over internal APIs
+
+Cons
+
+- need to maintain almost identical protobuf schema to relay requests to internal services
+- each resolver sends a separate call to proxied API (although it's worth noting that thanks to gRPC connection pool active connection can be maintained making the cost of remote calls not that high)
+
+### 2. Multiple schemas stitched into one
+
+#### 2.1. Apollo Server
+
+The simplest way to stitch multiple GraphQL schemas would be to use the Node.js [Apollo Server](https://www.apollographql.com/), that already has the functionality to merge schemas and proxy traffic to specific internal graphql servers. Merging works by introspection of existing schemas, and it takes care of things such as type conflicts. It is even possible to combine and modify types and fields from different schemas. Apollo supports queries, mutations, and subscriptions.
+
+Pros
+
+- easy to configure
+- hosts fully functioning GraphQL server, so that API user can't tell he's using an abstraction over few separate APIs
+- out of the box support for proxying queries, mutations, and subscriptions
+- type conflict detection and resolution
+- combining and editing types of merged schemas
+- single endpoint
+- a single call to proxied API
+
+Cons
+
+- adding Node.js to our technology stack (worse performance)
+- no facade over internal APIs
+
+#### 2.2. Custom HTTP proxy in Go
+
+Another idea was to implement an HTTP proxy that would first create a mapping of queries and remote servers they should access. Then it would forward received queries and mutations to remote servers using the mapped relations.
+
+That would, however, require us to give up on features such as introspection because they would become hidden behind the proxy without hosting the merged schema on a real GraphQL server.
+
+Handling subscriptions would be problematic as well because that would require proxying WebSocket traffic.
+
+Pros
+
+- avoiding Node.js (better performance)
+- single endpoint
+- a single call to proxied API
+
+Cons
+
+- lack of some GraphQL features (introspection)
+- implementing subscriptions proxying
```

Comment:
implementing subscriptions proxying: if we have Gateway in Go we always have to implement subscription proxying which makes me sad 😞